### ISSUE TITLE
Prometheus stopped scrapping metrics (urgency score 1)

### ISSUE THREAD

**Issue by galexrt** [coreos/prometheus-operator#326]
**What did you do?**
Run Prometheus for some time.

**What did you expect to see?**
Prometheus keep on scrapping data.

**What did you see instead? Under which circumstances?**
No data scrapped anymore.

**Environment**

* Kubernetes version information:
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:33:11Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-03-09T11:55:06Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 
Kubernetes contrib Ansible.

* Manifests:

```
apiVersion: v1   
kind: Pod
metadata:
  annotations:
    kubernetes.io/created-by: |
      {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"StatefulSet","namespace":"monitoring","name":"prometheus-main","uid":"880b847c-2a01-11e7-a743-c86000df09e0","apiVersion":"apps","resourceVersion":"4813488"}}
    pod.beta.kubernetes.io/hostname: prometheus-main-0
    pod.beta.kubernetes.io/subdomain: prometheus-operated
  creationTimestamp: 2017-04-25T21:52:52Z
  generateName: prometheus-main-
  labels:
    app: prometheus
    prometheus: main
  name: prometheus-main-0
  namespace: monitoring
  resourceVersion: "4813531"
  selfLink: /api/v1/namespaces/monitoring/pods/prometheus-main-0
  uid: 881e8ca2-2a01-11e7-a95e-c86000e0d5c2
spec:
  containers:
  - args:
    - -storage.local.retention=744h
    - -storage.local.num-fingerprint-mutexes=4096
    - -storage.local.path=/var/prometheus/data
    - -storage.local.chunk-encoding-version=2
    - -config.file=/etc/prometheus/config/prometheus.yaml
    - -storage.local.target-heap-size=8589934592
    image: quay.io/prometheus/prometheus:v1.6.1
    imagePullPolicy: IfNotPresent
    name: prometheus
    ports:
    - containerPort: 9090
      name: web
      protocol: TCP
    readinessProbe:
      failureThreshold: 100
      httpGet:
        path: /status
        port: web
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 3
    resources:
      limits:
        cpu: 2500m
        memory: 16Gi
      requests:
        cpu: 2500m
        memory: 12Gi
    terminationMessagePath: /dev/termination-log
    volumeMounts:
    - mountPath: /etc/prometheus/config
      name: config
      readOnly: true
    - mountPath: /etc/prometheus/rules
      name: rules
      readOnly: true
    - mountPath: /var/prometheus/data
      name: prometheus-main-db
      subPath: prometheus-db
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-6tk2d
      readOnly: true
  - args:
    - -reload-url=http://localhost:9090/-/reload
    - -config-volume-dir=/etc/prometheus/config
    - -rule-volume-dir=/etc/prometheus/rules
    image: quay.io/coreos/prometheus-config-reloader:v0.0.1
    imagePullPolicy: IfNotPresent
    name: prometheus-config-reloader
    resources:
      limits:
        cpu: 5m
        memory: 10Mi
      requests:
        cpu: 5m
        memory: 10Mi
    terminationMessagePath: /dev/termination-log
    volumeMounts:
    - mountPath: /etc/prometheus/config
      name: config
      readOnly: true
    - mountPath: /etc/prometheus/rules
      name: rules
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-6tk2d
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: game1-taker1.ger.htz.clstersys.net
  restartPolicy: Always
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 600
  volumes:
  - name: prometheus-main-db
    persistentVolumeClaim:
      claimName: prometheus-main-db-prometheus-main-0
  - name: config
    secret:
      defaultMode: 420
      secretName: prometheus-main
  - emptyDir: {}
    name: rules
  - name: default-token-6tk2d
    secret:
      defaultMode: 420
      secretName: default-token-6tk2d
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2017-04-25T21:52:52Z
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2017-04-25T21:53:02Z
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2017-04-25T21:52:52Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://3965c0958a87cf1eeb895f338d35dd648f646ac5bb051316d423dc968b682e61
    image: quay.io/prometheus/prometheus:v1.6.1
    imageID: docker-pullable://quay.io/prometheus/prometheus@sha256:448406d0e3d136d71c2e4c396df2a075c618d4556af87314439130fb24822b8f
    lastState: {}
    name: prometheus
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: 2017-04-25T21:52:56Z
  - containerID: docker://f60461e1fa6a240927d5b6b5d78c9f443bc9e294538767599c3d4a1721da9d1d
    image: quay.io/coreos/prometheus-config-reloader:v0.0.1
    imageID: docker-pullable://quay.io/coreos/prometheus-config-reloader@sha256:2c828783482b299882f21bd67b3dfbfdc7a7499aac3af2a812390079c0de8e35
    lastState: {}
    name: prometheus-config-reloader
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: 2017-04-25T21:52:57Z
  hostIP: 1.1.1.1
  phase: Running
  podIP: 100.64.114.4
  startTime: 2017-04-25T21:52:52Z
```

***

Is this related to my earlier issue with the Prometheus heap size?
I attached a graph of the urgency score. I didn't change the configuration, no new pods or services (ServiceMonitors, etc.) were added to the cluster.
![2017-05-02-125500_1600x880_scrot](https://cloud.githubusercontent.com/assets/3718398/25615448/7e584b20-2f37-11e7-80c0-1a7d3611bee2.png)
Why did the urgency score begin to go up there? Is there a way to tell why?
May it be caused by the underlying storage?

**Comment by fabxc** [comment]
Yes, that's a storage metric indicating load being too high for the available resources.
What resource limits does the Prometheus TPR have configured? How many targets does it scrape and how many samples per second?

**Comment by galexrt** [comment]
The Prometheus TPR manifest:
```
apiVersion: monitoring.coreos.com/v1alpha1
kind: Prometheus
metadata:
  labels:
    app: prometheus
    prometheus: main
  name: main
  namespace: monitoring
spec:
  alerting:
    alertmanagers:
    - name: alertmanager-main
      namespace: monitoring
      port: web
  replicas: 2
  resources:
    limits:
      cpu: "2.5"
      memory: "16Gi"
    requests:
      memory: "12Gi"
  ruleSelector:
    matchLabels:
      role: prometheus-main-rulefiles
      prometheus: main
  serviceMonitorSelector:
    matchExpressions:
    - key: app
      operator: In
      values:
      - node-exporter
      - ceph-exporter
      - mysql-exporter
      - prometheus
      - alertmanager
      - nginx-exporter
      - nginx-ingress-exporter
      - kube-state-metrics
      - kube-apiserver
      - k8s-apps-http
      - k8s-apps-https
  version: v1.6.1
  retention: 744h
  storage:
    class: rbd
    resources:
      requests:
        storage: 80Gi
```
The Prometheus instance(s) are scraping ~24 targets.
What is the metric name for samples per second? I assume it is `prometheus_local_storage_ingested_samples_total` here, after about two hours running of a fresh prometheus instance (new pvcs), I have reached `prometheus_local_storage_ingested_samples_total` ~28.092.413 samples.

This problem only occured with Prometheus 1.6.x. With Prometheus v1.5.x it worked on the same cluster without an issue.

**Comment by geoand** [comment]
Hello,

I am using the Prometheus Operator v0.8.2 with Prometheus v1.6.1 on a Kubernetes v1.6.2 cluster, and I am seeing the same exact issue (Prometheus entering rushed mode after a certain amount of time).

It should be noted that the necessary Kubernetes manifests were deployed using the `contrib/kube-prometheus/hack/cluster-monitoring/deploy` script (albeit changed to use only 1 replica for Prometheus and Alertmanager, and also with altered memory settings as explained in the next sentence).

Furthermore the duration which the monitoring actually works before entering rushed mode is correlated to the memory request/limits declared in the Prometheus manifest. The latest memory settings I used were
```
  resources:
    requests:
      memory: "3Gi"
    limits:
      cpu: "2.5"
      memory: "6Gi"
```

for which the cluster monitoring worked for about 24 hours.

Any direction on what actions I should take would be greatly appreciated!

Thanks,
George

**Comment by brancz** [comment]
@geoand if you can tell us how many targets and time-series you are talking about we can give you an estimate on required resources. Your Prometheus resources pretty much scale with those two factors. Note: 3-6Gi are not a whole lot for a real time time series database so my guess is that that is just too small for the number of targets and time-series you are monitoring.

**Comment by geoand** [comment]
@brancz Thank you the immediate reply!

I simply used the manifests that are provided in the `contrib/kube-prometheus/hack/cluster-monitoring` in order to monitor a Cluster of 5 nodes without adding any of my own service monitors.
I should also note that currently the cluster is receiving very little traffic since I am still evaluating it.

Is there any more information that I could provide?
Moreover, is there any documentation on how to estimate the resources that are needed for monitoring infrastructure depending on the number of targets and time-series?

Thanks again!

**Comment by brancz** [comment]
@geoand No problem :slightly_smiling_face: !

Here is the offical documentation on how to scale your deployment memory wise: https://prometheus.io/docs/operating/storage/#memory-usage. Note that this is one of the things we are trying to manage with the Prometheus Operator so we [automatically adjust those flags](https://github.com/coreos/prometheus-operator/blob/master/pkg/prometheus/statefulset.go#L267-L270) based on the resource requests, which is a pretty simple formula, just 2/3 of the memory request is used as the target heap size.

Further readings:

* https://prometheus.io/docs/operating/storage/#settings-for-high-numbers-of-time-series
* https://prometheus.io/docs/operating/storage/#helpful-metrics

TL;DR if your Prometheus instance often enters and leaves rushed mode either increase the memory or reduce the number of time-series being ingested (for example by not scraping targets that expose metrics you are not interested in).

**Comment by geoand** [comment]
@brancz Great information, thank you very much!

**Comment by galexrt** [comment]
@brancz Thanks for the informations!
But why do the same resource limits work for Prometheus 1.5.x and now with Prometheus 1.6.x this appears?

**Comment by brancz** [comment]
That could be because the ration we chose in 1.5.x was a lot more pessimistic than what we hoped to get out of 1.6.x. We may need to be a bit more pessimistic with our formula with 1.6.x, possibly 2/3 of the requested memory is too optimistic.

We actually also had an incident where Prometheus stopped ingesting data for no apparent reason, we are still investigating what happened there, however, it does not look like the same happened as it did here.

**Comment by xetys** [comment]
Hello. Like @geoand I have exactly the same issue. I also used the contrib scripts for cluster monitoring. My setup spans 9 nodes and with default 400Mi for the prometheus is giving me a working time of approximately 20 minutes. Scaling up to 2Gi increases that to 13 - 20 hours. After that, it slowly drops the scrape ratio, until it stops scrapping at all.

If looking to suggestions earlier this post, the way to go is to set the memory limit higher. Currently this looks not as a solution, but a postponing of the problem

**Comment by brancz** [comment]
We are particularly seeing this with Prometheus 1.6. I'd recommend staying with 1.5 for now, it will also use a lot of memory but we are not seeing the unstable scraping behavior.

**Comment by gianrubio** [comment]
Same issue on high memory consumption here, I'm on prom 1.6 and operator 0.9.0. 
Fixing the memory `limit` help to the k8s node stability but sometimes prometheus restart, it's acceptable for now.

**Comment by fabxc** [comment]
We are still investigating. It definitely seems to come from the Kubernetes client library being used in the service discovery. Oddly enough, the memory usage doesn't show in our typical prombench setup.

We are running an instance with an upgraded client library and deduplicating API watches across scrape configuration. The memory usage is significantly lower. It is just not clear yet which of the changes caused the drop. We are also seeing crashes, which are hopefully just a bug in our changes.

**Comment by geoand** [comment]
I should note that using the same setup I mentioned in the my previous comment and after downgrading to Prometheus 1.5.x, the monitoring infrastructure is working fine.

**Comment by fabxc** [comment]
It seems like a simple update of the k8s client libraries in Prometheus solves it. PR here: prometheus/prometheus#2704

We'll double check again in an hour and then do patch releases for 1.5 and 1.6

**Comment by geoand** [comment]
Sounds great!

Once the release it ready, I will immediately use it and post my experience here.

**Comment by brancz** [comment]
1.5.3 and 1.6.2 have been released :)

**Comment by geoand** [comment]
Wonderful! I will switch to `1.6.3` and report back in a few hours

**Comment by gianrubio** [comment]
It's planned a new release for prometheus-operator? As I notice it keep changing the image on the statefulset prometheus-k8s so it's impossible to setup the prom version.

**Comment by xetys** [comment]
you can just tell the prometheus version in the Prometheus config. See

```
apiVersion: monitoring.coreos.com/v1alpha1
kind: Prometheus
metadata:
  name: k8s
  labels:
    prometheus: k8s
spec:
  replicas: 2
  version: v1.5.0
  # ...
```

in https://github.com/xetys/kismatic-prometheus/blob/master/manifests/prometheus/prometheus-k8s.yaml

**Comment by gianrubio** [comment]
you're right @xetys, thanks!

**Comment by geoand** [comment]
I have been running the Kubernetes monitoring infrastructure with the new Prometheus `1.6.2` release for around 16 hours using a limit of 3GB of RAM (2GB of RAM for requested memory) and Prometheus has not entered the rushed state at all (whereas with `1.6.1` and my specific configuration of time series metrics it took only a few hours to enter rushed state).

If anything changes I will let you know.

Thanks!

**Comment by ntfrnzn** [comment]
Also looking good for me.  I upgraded my two test k8s clusters (k8s v1.5.6 and k8s v1.6.2) to the latest prometheus v1.6.2 and they've been stable for 48 hours, no rushed state, prometheus_local_storage_persistence_urgency_score holding below 0.05.

**Comment by fabxc** [comment]
Great, looks like this is solved. Also saw much better behavior in another k8s setup myself.
Closing here then.
